witchcraft/internal/tcpjson/tcp_logger.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tcpjson
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"sync"
+
+	werror "github.com/palantir/witchcraft-go-error"
+	"github.com/rs/zerolog"
+)
+
+var _ io.Writer = (*TCPWriter)(nil)
+
+const errWriterClosed = "writer is closed"
+
+// envelopeSerializerFunc provides a way to change the serialization method for the provided payload.
+type envelopeSerializerFunc func(payload []byte) ([]byte, error)
+
+// TCPWriter writes logs to a TCP socket and wraps them with envelope metadata.
+type TCPWriter struct {
+	provider           ConnProvider
+	envelopeSerializer envelopeSerializerFunc
+
+	// closedChan is used to signal that the writer is shutting down
+	closedChan chan struct{}
+
+	mu   sync.RWMutex // guards conn below
+	conn net.Conn
+}
+
+func NewTCPWriter(metadata LogEnvelopeMetadata, provider ConnProvider) *TCPWriter {
+	return newTCPWriterInternal(provider, zerologSerializer(metadata))
+}
+
+func newTCPWriterInternal(provider ConnProvider, serializerFunc envelopeSerializerFunc) *TCPWriter {
+	return &TCPWriter{
+		envelopeSerializer: serializerFunc,
+		provider:           provider,
+		closedChan:         make(chan struct{}),
+		conn:               nil,
+	}
+}
+
+func (d *TCPWriter) Write(p []byte) (int, error) {
```

Comment:
Think comment would be helpful for this as well: describe that input is wrapped, written to connection, behavior if there is a connection error etc.